bin/prepare_entity_binding_ratios.py

```diff
@@ -76,40 +109,72 @@ def main(args=None):
-
-        # entity-wise: do not account for entity-weight?
-        data = predictions[predictions.allele_id == allele_id] \
-                .merge(protein_peptide_occs) \
-                .merge(entities_proteins_occs) \
-                .drop(columns="protein_id") \
-                .merge(microbiomes_entities_occs) \
-                .merge(conditions) \
-                .drop(columns="microbiome_id") \
-                .merge(condition_allele_map) \
-                .drop(columns=["allele_id", "condition_id"])
-
-        n = len(data)
+    print("Prepare df with protein info ...", file = sys.stderr, flush=True, end='')
+    # Prepare df for joining protein information
+    protein_info = microbiomes_entities_occs \
+        .merge(conditions) \
+        .drop(columns="microbiome_id") \
+        .merge(condition_allele_map) \
+        .drop(columns="condition_id") \
+        .merge(entities_proteins_occs)
+    # -> protein_id, entity_id, entity_weight, condition_name, allele_id
+    # merged against condition_allele_map to keep only entities, and thus proteins, for which a prediction is requested for the current allele
+    print("\nInfo: protein_info", flush=True)
+    protein_info.info(verbose = False, memory_usage=print_mem)
+
+    # Process predictions chunk-wise based on peptide_ids
+    # (predictions chunk can contain more than chunk_size rows due to multiple alleles)
+    entity_results = pd.DataFrame()
+    max_peptide_id = predictions.index.max()
+    for i in range(0, max_peptide_id, args.chunk_size):
+        print("\nChunk peptide_ids: ", i, " - ", i+args.chunk_size-1)
+
+        now = datetime.datetime.now()
+        print ("Time: ...")
+        print (now.strftime("%Y-%m-%d %H:%M:%S"))
+
+        # Join predictions with protein_ids and further protein info
+        # NOTE would be faster to process directly respective chunks (see comment above)
+        data = predictions[(predictions.index >= i) & (predictions.index < i+args.chunk_size)] \
+                .join(protein_peptide_occs[(protein_peptide_occs.index >= i) & (protein_peptide_occs.index < i+args.chunk_size)]) \
+                .merge(protein_info)    # based on protein_id, allele_id
+        # -> index, prediction_score, allele_id, protein_id, count, entity_id, entity_weight, condition_name
+        # (protein_id could be dropped, but no big impact here)
+
+        # Call binder based on prediction_score
         data["binder"] = data["prediction_score"].apply(call_binder, method=args.method)
+        data.drop(columns="prediction_score", inplace=True)
 
+        # Count total number of peptides and number of binders for each entity, allele and condition (including multiple counts within proteins)
+        # Create extra column containing only counts for peptides that are classified as 'binder' to allow efficient aggregation
+        data["binder_count"] = 0
+        data.loc[data["binder"], "binder_count"] = data["count"]
```

Comment:
You could make it a one liner like this if I got it correctly.

`data["binder_count"] = [pep["count"] if pep["binder"] else 0 for _, pep in data.iterrows()]`

Reply:
True, but I think using `iterrows()` would be slower. But I don't find a comparison just now

Reply:
After benchmarking and research: Your solution takes 0.02s for 1m entries while mine took 18s! If changing the assignment in my solution to apply it takes 6s.

Reply:
The vectorized version turned out to be faster, thanks @tillenglert for comparison